nautobot/utilities/factory.py

```diff
@@ -13,14 +36,62 @@ class Meta:
                 exclude = ("has_group,")
 
             # Required foreign key
-            user = random_instance(User)
+            user = random_instance(User, allow_null=False)
 
             # Optional foreign key
             has_group = factory.Faker("pybool")
             group = factory.Maybe("has_group", random_instance(Group), None)
+
+            # Foreign key selected from a filtered queryset
+            tenant = random_instance(Tenant.objects.filter(group__isnull=False))
+
+            # Foreign key selected from a queryset generated by a lambda
+            # This needs to be done this way because .get_for_model() evaluates a ContentType queryset,
+            # and we need to defer evaluation of that queryset as well.
+            status = random_instance(lambda: Status.objects.get_for_model(Object), allow_null=False)
+    """
+
+    def get_random_instance():
+        queryset = _get_queryset_from_model_or_queryset_or_lambda(model_or_queryset_or_lambda)
+
+        if not allow_null and not queryset.exists():
+            raise RuntimeError(f"No objects in queryset for {model_or_queryset_or_lambda}! {queryset.explain()}")
+
+        return factory.random.randgen.choice(queryset) if queryset.exists() else None
+
+    return factory.LazyFunction(get_random_instance)
+
+
+def get_random_instances(model_or_queryset_or_lambda, maximum=None):
+    """
+    Factory helper - retrieve a random number of instances of the given model.
+
+    This is different from random_instance() in that it's not itself a lazy function generator, but should instead be
+    called only from within a @lazy_attribute or @post_generation function.
+
+    This is not an evenly weighted distribution (all counts equally likely), because in most of our code,
+    the relevant code paths distinguish between 0, 1, or >1 instances - there's not a functional difference between
+    "2 instances" and "10 instances" in most cases. Therefore, this implementation provides:
+        - 1/3 chance of no instances
+        - 1/3 chance of 1 instance
+        - 1/3 chance of (2 to n) instances, where each possibility is equally likely within this range
+
+    Args:
+        model_or_queryset_or_lambda: Either a model class, a model queryset, or a lambda that returns one of those
+        maximum (int): Maximum number of objects to return, or None for no limit
     """
-    return factory.LazyFunction(
-        lambda: factory.random.randgen.choice(model.objects.all()) if model.objects.count() else None
+    branch = factory.random.randgen.randint(0, 2)
+    queryset = _get_queryset_from_model_or_queryset_or_lambda(model_or_queryset_or_lambda)
+    count = queryset.count()
+    if maximum is None:
+        maximum = count
+    if branch == 0 or count == 0 or maximum == 0:
+        return []
+    if branch == 1 or count == 1 or maximum == 1:
```

Comment:
IMO `any()` is more readable in times like this.

```suggestion
    if not any([branch, count, maximum]):
        return []
    if any([branch == 1, count == 1, maximum == 1]):
```